drivers/rilmodem/radio-settings.c

```diff
@@ -87,16 +149,339 @@ void ril_set_rat_mode(struct ofono_radio_settings *rs,
+	struct reply_radio_capability *caps;
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: error %s", __func__,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL)
+		ofono_error("%s: parse error", __func__);
+
+	g_free(caps);
+}
+
+static void switch_start_cb(struct ril_msg *message, gpointer user_data)
+{
+	struct radio_data *rd = user_data;
+	struct switch_data *sd = rd->switch_d;
+	struct radio_data *rd1 = sd->rd_1;
+	struct radio_data *rd2 = sd->rd_2;
+	struct reply_radio_capability *caps;
+
+	sd->pending_msgs--;
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: error %s", __func__,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL) {
+		ofono_error("%s: parse error", __func__);
+		return;
+	}
+
+	if (sd->pending_msgs != 0)
+		return;
+
+	DBG("Sending requests for APPLY phase");
+
+	send_set_radio_cap(rd1, g_session, RIL_RC_PHASE_APPLY,
+				rd2->ril_rats, rd2->modem_uuid,
+				RIL_RC_STATUS_NONE, switch_apply_cb);
+	send_set_radio_cap(rd2, g_session, RIL_RC_PHASE_APPLY,
+				rd1->ril_rats, rd1->modem_uuid,
+				RIL_RC_STATUS_NONE, switch_apply_cb);
+	sd->pending_msgs = 2;
+
+	g_free(caps);
+}
+
+static void switch_caps(struct switch_data *sd)
+{
+	struct radio_data *rd1 = sd->rd_1;
+	struct radio_data *rd2 = sd->rd_2;
+
+	/* START phase */
+	g_session++;
+
+	send_set_radio_cap(rd1, g_session, RIL_RC_PHASE_START,
+				rd1->ril_rats, rd1->modem_uuid,
+				RIL_RC_STATUS_NONE, switch_start_cb);
+	send_set_radio_cap(rd2, g_session, RIL_RC_PHASE_START,
+				rd2->ril_rats, rd2->modem_uuid,
+				RIL_RC_STATUS_NONE, switch_start_cb);
+	sd->pending_msgs = 2;
+}
+
+static void get_rs_with_mode(struct ofono_modem *modem, void *data)
+{
+	struct switch_data *sd = data;
+	struct radio_data *rd_ref = sd->rd_1;
+	struct ofono_atom *atom;
+	struct ofono_radio_settings *rs;
+	struct radio_data *rd;
+	const char *standby_group, *modem_group;
+
+	atom = __ofono_modem_find_atom(modem, OFONO_ATOM_TYPE_RADIO_SETTINGS);
+	if (atom == NULL)
+		return;
+
+	rs = __ofono_atom_get_data(atom);
+	rd = ofono_radio_settings_get_data(rs);
+	if (rd == rd_ref)
+		return;
+
+	standby_group = ofono_modem_get_string(rd_ref->modem, "StandbyGroup");
+	if (standby_group == NULL)
+		return;
+
+	modem_group = ofono_modem_get_string(modem, "StandbyGroup");
+	if (g_strcmp0(standby_group, modem_group) != 0)
+		return;
+
+	if ((rd->available_rats & sd->mode_to_switch) == 0)
+		return;
+
+	sd->rd_2 = rd;
+}
+
+void ril_set_rat_mode(struct ofono_radio_settings *rs,
+			enum ofono_radio_access_mode mode,
+			ofono_radio_settings_rat_mode_set_cb_t cb,
+			void *data)
+{
+	struct radio_data *rd = ofono_radio_settings_get_data(rs);
+	struct cb_data *cbd = cb_data_new(cb, data, rs);
+	struct switch_data *sd = NULL;
+
+	if (rd->switch_d != NULL)
+		goto error;
```

Comment:
Nope, this block of code is a called a driver atom.  It's created by a probe() call and destroyed by a remove() call ( see the associated driver table at the bottom of the file ).  When probe() is called, it creates a local data structure called radio_data, and it stores a pointer to this structure in the core ofono radio_settings instance.  So ofono_radio_settings_get_data should always return a valid pointer.